### PR TITLE
flesh out signature validation

### PR DESCRIPTION
**Pull request by wraithgar**
allow secret to be obtained via callback function
allow url to be a wildcard

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -74,85 +74,91 @@ function serverHandler(req, res) {
...
-            }
 
-            signature = signature.replace(/^sha1=/, '');
-            var digest = Crypto.createHmac('sha1', self.secret).update(data).digest('hex');
-
-            if (signature !== digest) {
-                self.logger.error('got invalid signature, returning 403');
+        self.getSecret(req, function (err, secret) {
+            if (err) {
+                this.logger.error('error getting secret for' +  res.url + 'returning 403');
```
Comment (nlf) [inline index.js]: this should be self

File: index.js [hunk 2]
```diff
@@ -74,85 +74,91 @@ function serverHandler(req, res) {
...
-        if (isForm) {
-            data = Querystring.parse(data).payload;
-        }
-        data = parse(data);
+            var signature = req.headers['x-hub-signature'];
+            if (secret) {
 
-        var event = req.headers['x-github-event'] || (req.headers['x-gitlab-event'] ? req.headers['x-gitlab-event'].split(' ')[0].toLowerCase() : 'unknown');
+                if (!signature) {
+                    this.logger.error('secret configured, but missing signature, returning 403');
```
Comment (nlf) [inline index.js]: here too

File: index.js [hunk 3]
```diff
@@ -74,85 +74,91 @@ function serverHandler(req, res) {
...
+                    return reply(403, res);
+                }
 
-        // invalid json
-        if (!data) {
-            self.logger.error(Util.format('received invalid data from %s, returning 400', remoteAddress));
-            return reply(400, res);
-        }
+                signature = signature.replace(/^sha1=/, '');
+                var digest = Crypto.createHmac('sha1', this.secret).update(data).digest('hex');
```
Comment (nlf) [inline index.js]: also here

File: index.js [hunk 4]
```diff
@@ -74,85 +74,91 @@ function serverHandler(req, res) {
...
 
-        // handle GitLab system hook
-        if (event !== 'system'){
-            // invalid json
-            if (!data.repository || !data.repository.name) {
-                self.logger.error(Util.format('received incomplete data from %s, returning 400', remoteAddress));
-                return reply(400, res);
+                if (signature !== digest) {
+                    this.logger.error('got invalid signature, returning 403');
```
Comment (nlf) [inline index.js]: and here

File: index.js (hunk 1 above)
Comment (wraithgar) [inline index.js]: thanks, it went through a few iterations before everything came back into the same function.

**Pull request merged**